design/keycloak.x/configuration.md

```diff
@@ -0,0 +1,159 @@
+# Keycloak.X Server Configuration
+
+* **Status**: Draft #1
+
+## How to set configuration
+
+Keycloak.X can be configured in a property file, using command-line arguments or environment variables. All configuration
+properties are available in all different approaches.
+
+Properties (conf/keycloak.properties):
+
+    hostname.frontendUrl=https://mykeycloak
+    
+Environment variables:
+
+    export KC_HOSTNAME_FRONTEND-URL=https://mykeycloak
+    bin/kc.sh
+
+Command-line arguments:
+
+    bin/kc.sh --hostname-frontendUrl=https://mykeycloak
+    
+When evaluating configuration if a property is defined in several places the order of resolution is first in the list of:
+command-line arguments, environment variables, properties.  
+
+The format used in properties will be used throughout the documentation.
+
+The keys used above follows a strict rewrite rule, where the key in properties is the name of the property. Property 
+names should follow the following rules:
+
+* Use camel-case (`propertyName`)
+* Use dot (`.`) for separation
+* All options must be in a category (`<categoryName>.<propertyName>`)
+* Multiple levels of categories are permitted (`<categoryName>.<subCategoryName>.<propertyName>`)
+
+Rules for converting property keys to different formats follow these rules:
+
+* Environment variables: 
+  - Prefix with `KC_`
+  - Characters converted to uppercase
+  - Dot (`.`) replaced with underscore (`_`)
+  - Use hyphen (`-`) for camel-case names (`PROPERTY-NAME`)
+* Command-line arguments:
+  - Dot (`.`) converted to hyphen (`-`)
+  
+## SPIs and Provider Configuration
+
+Properties for providers follow the following format:
+
+    spi.<spi-name>.<provider-name>.<config-name>
+
+For example:
+
+    spi.hostname.default.frontendUrl=https://mykeycloak
+    
+## Alias
+
+It should not be required to use SPI/provider specific property names for commonly configured properties for default
+providers. Here we will support alias for properties.
```

Comment:
Is that to remain compatible with the existing Keycloak SPI provider configuration?

Reply:
Not really. We should have alias/high-level properties for common configuration things, then have the spi/provider specific properties for less common things, and custom providers (custom providers shouldn't use high-level/alias as that would then easily end up with conflicting with KC specific ones).